### PR TITLE
Virtual columns for Kafka headers

### LINKED ISSUE #9337: Kafka headers
Hi! We intend to use the Kafka Engine (https://clickhouse.tech/docs/ru/operations/table_engines/kafka/) for storing events in Clickhouse.
Also, some of the necessary information is stored in the Kafka's headers (https://issues.apache.org/jira/browse/KAFKA-4208) and we would like to store them too.
Is there a way to achive this ?

### PR DESCRIPTION
**Pull request by filimonov**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- New Feature

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Add 2 more virtual columns for engine=Kafka to access message headers.

Detailed description / Documentation draft:
https://cwiki.apache.org/confluence/display/KAFKA/KIP-82+-+Add+Record+Headers

Resolves #9337
```
_headers.name Array(String)
_headers.value Array(String)
```

Not implementing that for Producer right now (it would a bit too stupid to write them both to datastream and to headers). Maybe after that: https://github.com/ClickHouse/ClickHouse/issues/9436 

PR is on the top of https://github.com/ClickHouse/ClickHouse/pull/11260

### CODE REVIEW

**Comment by filimonov** [comment]
2 concerns:
1) optionate filling of those columns? 
2) use `Array(LowCardinality(String))` for header names?

File: src/Storages/Kafka/KafkaBlockInputStream.cpp [hunk 1]
```diff
@@ -19,9 +19,8 @@ KafkaBlockInputStream::KafkaBlockInputStream(
     , column_names(columns)
     , max_block_size(max_block_size_)
     , commit_in_suffix(commit_in_suffix_)
-    , non_virtual_header(storage.getSampleBlockNonMaterialized()) /// FIXME: add materialized columns support
-    , virtual_header(storage.getSampleBlockForColumns({"_topic", "_key", "_offset", "_partition", "_timestamp"}))
-
+    , non_virtual_header(storage.getSampleBlockNonMaterialized())
+    , virtual_header(storage.getSampleBlockForColumns({"_topic", "_key", "_offset", "_partition", "_timestamp","_timestamp_ms","_headers.name","_headers.value"}))
```
Comment (alesapin) [inline src/Storages/Kafka/KafkaBlockInputStream.cpp:23]: I think we should also add them to the storage virtuals list https://github.com/clickhouse/ClickHouse/blob/master/src/Storages/Kafka/StorageKafka.cpp#L720-L729.

**Review by alesapin — CHANGES_REQUESTED** [review]
Just check storage virtuals. If it's OK, than PR LGTM.

File: src/Storages/Kafka/StorageKafka.cpp [hunk 2]
```diff
@@ -724,7 +726,10 @@ NamesAndTypesList StorageKafka::getVirtuals() const
         {"_key", std::make_shared<DataTypeString>()},
         {"_offset", std::make_shared<DataTypeUInt64>()},
         {"_partition", std::make_shared<DataTypeUInt64>()},
-        {"_timestamp", std::make_shared<DataTypeNullable>(std::make_shared<DataTypeDateTime>())}
+        {"_timestamp", std::make_shared<DataTypeNullable>(std::make_shared<DataTypeDateTime>())},
+        {"_timestamp_ms", std::make_shared<DataTypeNullable>(std::make_shared<DataTypeDateTime64>(3))},
+        {"_headers.name", std::make_shared<DataTypeArray>(std::make_shared<DataTypeString>())},
+        {"_headers.value", std::make_shared<DataTypeArray>(std::make_shared<DataTypeString>())}
```
Comment (filimonov) [inline src/Storages/Kafka/StorageKafka.cpp:732]: @alesapin ^^^

Comment (alesapin) [reply src/Storages/Kafka/StorageKafka.cpp:732]: Really)

**Review by alesapin — APPROVED** [review]

**Pull request merged**